### DISCUSSION TITLE
Confused about frontend hosting and pocketbase routing

### DISCUSSION THREAD

**Discussion by dpatel20** [pocketbase/pocketbase#6165] [Q&A] [answered]
I'm beginning to look at using pocketbase for a hobby project but I am little confused about routing.

I see that pocketbase can do routing but I am unclear how this works.

If I am making a web app and hosting the frontend on netlify or vercel, how does the pocketbase routing work?

So, if user goes to someapp.netlify.app/michael/ where 'michael' is {name} param in a pocketbase route, I don't understand how this gets to the pocketbase router. Wouldn't netlify say that it doesn't recognise this path rather than allow pocketbase router to handle it?

I am not a developer - more of a hobbyist - so apologies for the newbie/dumb questions!

(PS: was thinking of using pockethost for pocketbase hosting)

**Comment by thangngoc89** [comment]
If you're hosting on netlify then you probably don't need to worry about pocketbase's routing. You just run pocketbase somewhere and point your API to the pocketbase instance.

**Comment by dpatel20** [comment]
I recently found that pocketbase can also serve your website static files by including them in pb_public.

So, if I have, for example, a Svelte-based app then I can host that in pb_public folder on, say, pockethost.io and make use of the pocketbase router.

If I host the frontend on netlify, then I cannot use pocketbase router (and would use something on the netlify side like express).

Is the above understanding correct?

**Reply by thangngoc89** [reply]
@dpatel20 if you put the SPA (built using Svelte) to pocketbase, then pocketbase would show it using SPA mode. Meaning all requests (except admin and api) would display the `index.html` file. It’s up to your Svelte app to decide how to do routing client side.

**Reply by dpatel20** [reply]
Sorry but I am even more confused now...I don't understand what 'SPA mode' means for pocketbase - are you saying that pocketbase routing would not work and all routing would need to be handled by the SPA framework (e.g. Svelte) router? So, if a user went to _myapp.pockethost.io/michael_ then pocketbase would still just serve the SPA _index.html_ (not give a 404 error as _/michael_ page does not exist) and then the framework router would handle the _/michael_ part?

In that case, can you please explain when/how the pocketbase router would work/be invoked (since it does not get invoked if hosting frontend elsewhere (e.g. on netlify) and not in 'SPA mode')?

Thanks!

**Reply by thangngoc89** [reply]
@dpatel20 So in the purest form, framework like Svelte (I'm not familiar with this) or React or Vue works by making an index.html and add some Javascript inside. That's why it's called Single Page Application (SPA). 

The documentation of [sveltekit in spa mode](https://svelte.dev/docs/kit/single-page-apps) has an example of running on Apache you might want to check that out to understand more.

To answer your questions:

> So, if a user went to myapp.pockethost.io/michael then pocketbase would still just serve the SPA index.html (not give a 404 error as /michael page does not exist) and then the framework router would handle the /michael part?

This is correct. The SPA would handle it client side (parsing the path from browser and do client side routing)

> In that case, can you please explain when/how the pocketbase router would work/be invoked (since it does not get invoked if hosting frontend elsewhere (e.g. on netlify) and not in 'SPA mode')?

In not SPA mode, pocketbase acts as an API service and you don't have to care about it at all.

Let's say you start pocketbase at myapp.pockethost.io and you host your frontend in myapp.netlify.com. What you need to do is call pocketbase API at myapp.pockethost.io/api .

Hope this make senses to you.

You could search tutorials about Firebase, a very popular service that acts as Backend as a service (BaaS) just like pocketbase. There are probably more information to talk about the concept.

**Reply by dpatel20** [reply]
That's great, thanks. I think I need to take a look at a few tutorials and maybe just have a go!

But, hopefully one last clarification: what use case would you actually be able to use the pocketbase router?

**Reply by thangngoc89** [reply]
> what use case would you actually be able to use the pocketbase router?

Not sure why are you confused about this "pocketbase router" thing! Pocketbase would happily server static contents that you put to pb_public just like any static server. Or you may extend the "router" using Javascript or Go.

**Comment by sksar** [comment, marked as answer]
@dpatel20 


## pb_public folder

Consider the pb_public folder as a basic, fully static web hosting service akin to Firebase Hosting. It functions as a repository for storing your fully static HTML, CSS, and JavaScript files in a S3-like bucket and hosting them.

Any content placed within the pb_public folder is unrelated to any framework-based routing, such as SvelteKit or NextJS.

If you are constructing a SvelteKit website with a router, that router is distinct from the Pocketbase router. Since the pb_public serves solely as a static file hosting folder, it can only accommodate static files.

Therefore, when building a SvelteKit or NextJS project, there are two primary approaches: constructing it as a fully static website or incorporating server-rendered functions, such as server routes in Svelte or Next. However, if you utilize SvelteKit’s server routes, you essentially execute NodeJS on the server side. Platforms like Vercel and Netlify handle these aspects, hosting static files in a static file hosting service and serverless functions like AWS Lambda for dynamic server components.

In contrast, Pocketbase operates differently. The pb_public folder serves solely as a fully static hosting service, devoid of any server-side processing capabilities. It cannot execute any code, including NodeJS. Consequently, it is not suitable for hosting server-side code from your frontend framework.

It is important to note that SvelteKit and NextJS are full-stack frameworks, encompassing both client-side and server-side code. These frameworks enable the development of both frontend and backend code within a single codebase. When referring to a Single Page Application (SPA), we mean a website that loads content dynamically on a single page. In such cases, we utilize these frameworks without writing any backend code. The framework generates plain HTML, CSS, and JavaScript files for static hosting.

## Pocketbase router

The Pocketbase router presents a distinct experience. Imagine serverless functions encapsulated within the Pocketbase binary. Unlike NodeJS (if JavaScript is used), Pocketbase allows you to create custom endpoints, such as /api/abcd.
This functionality is unrelated to the frontend and is solely a Pocketbase backend concern. It has no connection to any framework’s router.

Utilizing Pocketbase hooks (.pb.js files), you can create your own API routes and essentially extend Pocketbase with custom backend logic. Pocketbase executes these .pb.js files whenever that API route is invoked. (This is akin to the experience you gain when starting with a framework like ExpressJS.)

These hooks enable you to extend Pocketbase with your own business logic and serve various purposes, such as performing computations before sending information to the client or implementing server-side logic that you wish to keep confidential.


<hr>


Feature | pb_public folder | Pocketbase router
-- | -- | --
Purpose | Static file hosting (S3-like bucket) | Serverless functions (encapsulated in Pocketbase binary)
Relationship to frontend frameworks | Unrelated to framework-based routing (e.g., SvelteKit, NextJS) | Distinct from frontend framework router
Code execution | Cannot execute code (e.g., NodeJS) | Can execute code through Pocketbase hooks (.pb.js files)
Suitability for server-side code | Not suitable | Suitable for creating custom endpoints and server-side logic
Examples of use | Hosting fully static HTML, CSS, and JavaScript files | Creating custom API routes, implementing server-side logic, performing computations before sending data to the client


Although my explanation may be lengthy, I trust it will dispel any doubts you may have.

**Comment by dpatel20** [comment]
Thanks @thangngoc89 and @sksar for taking the time to answer my questions!
